src/coreclr/interpreter/compiler.cpp

```diff
@@ -4612,37 +4911,643 @@ void InterpCompiler::EmitCall(CORINFO_RESOLVED_TOKEN* pConstrainedToken, bool re
+        execContextOffset = currentOffset;
+        INTERP_DUMP("Allocate ExecutableContext at offset %d\n", currentOffset);
+        int32_t slotIndex = currentOffset / sizeof(void*);
+        SetSlotToTrue(objRefSlots, slotIndex);
+        currentOffset += sizeof(void*);
+    }
+
+    int32_t keepAliveOffset = 0;
+    if (needsKeepAlive)
+    {
+        // Mark keep-alive pointer as a GC reference
+        keepAliveOffset = currentOffset;
+        INTERP_DUMP("Allocate KeepAlive at offset %d\n", currentOffset);
+        int32_t slotIndex = currentOffset / sizeof(void*);
+        SetSlotToTrue(objRefSlots, slotIndex);
+        currentOffset += sizeof(void*);
+    }
+
+    // Step 5: Get continuation type handle
+    assert(currentOffset / sizeof(void*) <= objRefSlots.GetSize());
+    CORINFO_CLASS_HANDLE continuationTypeHnd = m_compHnd->getContinuationType(
+        currentOffset,
+        objRefSlots.GetUnderlyingArray(),
+        (currentOffset / sizeof(void*)) * sizeof(bool)
+    );
+
+    // Step 6: Build InterpAsyncSuspendData structure
+    if (m_asyncResumeFuncPtr == NULL)
+    {
+        m_compHnd->getAsyncResumptionStub(&m_asyncResumeFuncPtr);
+        assert(m_asyncResumeFuncPtr != NULL);
+    }
+
+    InterpAsyncSuspendData* suspendData = (InterpAsyncSuspendData*)AllocMethodData(sizeof(InterpAsyncSuspendData));
+    m_asyncSuspendDataItems.Add(suspendData);
+    CORINFO_ASYNC_INFO asyncInfo;
+    m_compHnd->getAsyncInfo(&asyncInfo);
+    
+    GetDataForHelperFtn(CORINFO_HELP_ALLOC_CONTINUATION);
+    suspendData->continuationTypeHnd = continuationTypeHnd;
+    AllocateIntervalMapData_ForVars(&suspendData->liveLocalsIntervals, liveVars);
+    AllocateIntervalMapData_ForVars(&suspendData->zeroedLocalsIntervals, varsToZero);
+
+    int32_t flags = 0;
+    if (returnValueVar != -1)
+    {
+        flags |= CORINFO_CONTINUATION_HAS_RESULT;
+    }
+
+    if (captureContinuationContext)
+    {
+        flags |= CORINFO_CONTINUATION_HAS_CONTINUATION_CONTEXT;
+    }
+
+    if (ContinuationContextHandling == ContinuationContextHandling::ContinueOnThreadPool)
+    {
+        flags |= CORINFO_CONTINUATION_CONTINUE_ON_THREAD_POOL;
+    }
+
+    if (needsEHHandling)
+    {
+        flags |= CORINFO_CONTINUATION_HAS_EXCEPTION;
+    }
+
+    suspendData->flags = (CorInfoContinuationFlags)flags;
+
+    suspendData->offsetIntoContinuationTypeForExecutionContext = execContextOffset + OFFSETOF__CORINFO_Continuation__data;
+    suspendData->keepAliveOffset = keepAliveOffset + OFFSETOF__CORINFO_Continuation__data;
+    suspendData->captureSyncContextMethod = asyncInfo.captureContinuationContextMethHnd;
+    suspendData->restoreExecutionContextMethod = asyncInfo.restoreExecutionContextMethHnd;
+    suspendData->restoreContextsMethod = asyncInfo.restoreContextsMethHnd;
+    suspendData->resumeInfo.Resume = (size_t)m_asyncResumeFuncPtr;
+    suspendData->resumeInfo.DiagnosticIP = NULL;
+    // TODO! Register methodStartIP to be filled in as needed.
```

Comment:
Is this TODO still needed? We already set the methodStartIP in the UpdateWithFinalMethodByteCodeAddress, don't we?